src/i18n.ts

```diff
@@ -0,0 +1,342 @@
+import has from 'dojo-core/has';
+import global from 'dojo-core/global';
+import { Handle } from 'dojo-core/interfaces';
+import { assign } from 'dojo-core/lang';
+import Promise from 'dojo-shim/Promise';
+
+/**
+ * A default bundle used as basis for loading locale-specific bundles.
+ */
+export interface Bundle<T extends Messages> {
+	/**
+	 * The absolute module ID used by the loader to resolve paths to locale-specific paths. This MUST follow the format
+	 * "{basePath}{separator}{filename}". For example, if the module ID for a bundle is "nls/common", the loader will
+	 * look for locale-specific bundles at "nls/{locale}/common".
+	 */
+	baseUrl: string;
+
+	/**
+	 * A list of supported locales. Any included locale MUST have an associated bundle.
+	 */
+	locales?: string[];
+
+	/**
+	 * The map of default messages that will be used when locale-specific messages are unavailable.
+	 * Note that any message key used in the i18n system MUST have a default specified here.
+	 */
+	messages: T;
+}
+
+interface BundleMap<T extends Messages> {
+	[key: string]: T;
+}
+
+/**
+ * Represents an object from which a locale can be read. If a context object also has both `invalidate` and `own`
+ * methods, it will be updated (invalidated) when the locale is changed.
+ */
+export interface LocaleContext {
+	/**
+	 * The state object from which the locale string is read.
+	 */
+	state: { locale?: string };
```

Comment:
That would be more consistent with how state interfaces are defined elsewhere in Dojo 2.